virtcontainers/qemu.go

```diff
@@ -1275,6 +1275,59 @@ func (q *qemu) hotplugVhostUserDevice(vAttr *config.VhostUserDeviceAttrs, op ope
 	return nil
 }
 
+func (q *qemu) qomGetSlot(qomPath string) (uint8, error) {
+	addr, err := q.qmpMonitorCh.qmp.ExecQomGet(q.qmpMonitorCh.ctx, qomPath, "addr")
+	if err != nil {
+		return 0xff, err
+	}
+	addrf, ok := addr.(float64)
+	// XXX going via float makes no real sense, but that's how
+	// JSON works, and we should get away with it for the small
+	// values we have here
+	if !ok {
+		return 0xff, fmt.Errorf("addr QOM property of %s is %t not a number", qomPath, addr)
+	}
+	addri := uint8(addrf)
+
+	if (addri & 0x7) != 0 {
+		return 0xff, fmt.Errorf("Can't cope with non-zero function on %s", qomPath)
+	}
+
+	return addri >> 3, nil
+}
+
+func (q *qemu) vfioQuerySlots(qemuID string) (uint8, uint8, error) {
+	devSlot, err := q.qomGetSlot(qemuID)
+	if err != nil {
+		return 0xff, 0xff, err
+	}
+
+	busq, err := q.qmpMonitorCh.qmp.ExecQomGet(q.qmpMonitorCh.ctx, qemuID, "parent_bus")
+	if err != nil {
+		return 0xff, 0xff, err
+	}
+
+	bus, ok := busq.(string)
+	if !ok {
+		return 0xff, 0xff, fmt.Errorf("parent_bus QOM property of %s is %t not a string", qemuID, busq)
+	}
+
+	// Chop off the last QOM path component to go from the bus to
```

Comment:
Might be valuable to explain the structure of the qom path you receive, or give a pointer to some doc that does.